extension/lib/ActiveURIService.jsm

```diff
@@ -0,0 +1,149 @@
+    .getInterface(Ci.nsIDOMWindow)
+  );
+}
+
+this.ActiveURIService = {
+  focusedWindow: null,
+  focusedURI: null,
+
+  // Windows with registered listeners
+  trackedWindows: new Set(),
+
+  // Objects listening to changes in the active URI
+  observers: new Set(),
+
+  startup() {
+    Services.obs.addObserver(this, "xul-window-registered");
+    Services.obs.addObserver(this, "xul-window-destroyed");
+
+    // Register existing windows for tracking
+    const windowList = Services.wm.getEnumerator(null);
+    while (windowList.hasMoreElements()) {
+      this.trackWindow(windowList.getNext());
+    }
+
+    // Set the focused URI to the currently-focused URI if possible
+    const mostRecentWindow = Services.wm.getMostRecentWindow(null);
+    this.onFocusWindow(mostRecentWindow);
+  },
+
+  shutdown() {
+    Services.obs.removeObserver(this, "xul-window-registered");
+
+    // Clean up tracked windows
+    for (const domWindow of this.trackedWindows) {
+      this.untrackWindow(domWindow);
+    }
+  },
+
+  addObserver(observer) {
+    this.observers.add(observer);
+  },
+
+  removeObserver(observer) {
+    this.observers.delete(observer);
+  },
+
+  setFocusedURI(uri) {
+    this.focusedURI = uri;
+    for (const observer of this.observers) {
+      observer.observe(this, 'uriFocused', uri);
+    }
+  },
+
+  trackWindow(domWindow) {
+    this.trackedWindows.add(domWindow);
+    domWindow.addEventListener("focus", this);
+    domWindow.addEventListener("blur", this);
+    if (domWindow.gBrowser) {
+      domWindow.gBrowser.addProgressListener({
+        onLocationChange: this.onLocationChange.bind(this, domWindow),
+      });
+    }
+  },
+
+  untrackWindow(domWindow) {
+    domWindow.removeEventListener("focus", this);
+    domWindow.removeEventListener("blur", this);
+    if (domWindow.gBrowser) {
+      domWindow.gBrowser.removeProgressListener(this);
+    }
+    this.trackedWindows.delete(domWindow);
+  },
+
+  onRegisterWindow(domWindow) {
+    domWindow.addEventListener("load", () => {
+      this.trackWindow(domWindow);
+    });
+  },
+
+  onDestroyWindow(domWindow) {
+    if (this.trackedWindows.has(domWindow)) {
+      this.untrackWindow(domWindow);
+    }
+  },
+
+  onFocusWindow(domWindow) {
+    this.focusedWindow = domWindow;
+    if (domWindow.gBrowser) {
+      this.setFocusedURI(domWindow.gBrowser.currentURI);
+    } else {
+      this.setFocusedURI(null);
+    }
+  },
+
+  onBlurWindow(domWindow) {
+    if (domWindow === this.focusedWindow) {
+      this.focusedWindow = null;
+      this.setFocusedURI(null);
+    }
+  },
+
+  onLocationChange(domWindow, progress, request, uri) {
+    if (domWindow === this.focusedWindow) {
+      this.setFocusedURI(uri);
+    }
+  },
+
+  observe(subject, topic, data) {
+    switch (topic) {
+      case "xul-window-registered":
+        this.onRegisterWindow(getDOMWindow(subject));
+        break;
+      case "xul-window-destroyed":
+        this.onDestroyWindow(getDOMWindow(subject));
```

Comment:
This doesn't seem to do what we want.  `subject` here is null so we won't be able to destroy the window properly.  I added 63690d4051001b04c4af9ae7d644321cac3e6ef1 to prevent a memory leak in trackedWindows, but it's an ugly solution.